File: ts/game/creature/Creature.ts
import * as PIXI from 'pixi.js';
import { Vector } from '../Vector.js';
import { Camera } from '../Camera.js';
import { Level } from '../level/Level.js';

export class Creature {
  pos: Vector = Vector.zeroVector;
  vel: Vector = Vector.zeroVector;
  angle = 0;
  drawRadius = 0.5;
  hitRadius = 0.4;
  mass = 1;
  health: number;
  invincible = -0.01;
  permInvincible = false;
  solid = true;
  color?: [number, number, number];
  texture?: PIXI.Texture;
  display?: PIXI.DisplayObject;

  constructor(health: number, texOrColor: PIXI.Texture | [number, number, number]) {
    this.health = health;
    if (texOrColor instanceof PIXI.Texture) this.texture = texOrColor;
    else this.color = texOrColor;
  }

  update(dt: number, _level: Level) {
    this.vel = this.vel.add(this.vel.scale(-20 * dt));
    this.pos = this.pos.add(this.vel.scale(dt));
    while (this.angle >= 2 * Math.PI) this.angle -= 2 * Math.PI;
    while (this.angle <= -2 * Math.PI) this.angle += 2 * Math.PI;
    if (this.invincible >= 0) this.invincible -= dt;
    if (this.permInvincible) this.invincible = 0.01;
  }

  collide(other: Creature) {
    if (!this.solid || !other.solid) return;
    const dist = other.pos.sub(this.pos).magnitude();
    const depth = other.hitRadius + this.hitRadius - dist;
    if (depth >= 0) {
      const normal = other.pos.sub(this.pos).direction();
      const relVel = this.vel.sub(other.vel);
      const relVelR = relVel.dot(normal);
      const m1 = this.mass;
      const m2 = other.mass;
      const impact = (1 / (1 / m1 + 1 / m2)) * relVelR;
      this.vel = this.vel.add(normal.scale(-impact / m1));
      other.vel = other.vel.add(normal.scale(impact / m2));
      const disp = (1 / (1 / m1 + 1 / m2)) * depth;
      this.pos = this.pos.add(normal.scale(-disp / m1));
      other.pos = other.pos.add(normal.scale(disp / m2));
      this.onTouch(other);
      other.onTouch(this);
    }
  }

  onTouch(_other: Creature) {}

  takeHit(damage: number, dir: Vector) {
    if (this.invincible > 0) return;
    this.vel = this.vel.add(dir.direction().scale((1 / this.mass) * 20));
    if (damage <= 0) return;
    this.health -= damage;
    this.invincible = 0.7;
  }

  render(container: PIXI.Container) {
    if (this.texture) {
      if (!(this.display instanceof PIXI.Sprite)) {
        const spr = new PIXI.Sprite(this.texture);
        spr.anchor.set(0.5);
        spr.width = this.drawRadius * 2;
        spr.height = this.drawRadius * 2;
        spr.scale.y *= -1;
        this.display = spr;
        container.addChild(spr);
      }
      const spr = this.display as PIXI.Sprite;
      spr.position.set(this.pos.x, this.pos.y);
      spr.rotation = this.angle;
      spr.alpha = this.invincible > 0 ? 0.5 : 1;
    } else if (this.color) {
      if (!(this.display instanceof PIXI.Graphics)) {
        this.display = new PIXI.Graphics();
        container.addChild(this.display);
      }
      const g = this.display as PIXI.Graphics;
      g.clear();
      const col = (this.color[0] * 255 << 16) + (this.color[1] * 255 << 8) + (this.color[2] * 255);
      g.beginFill(col, this.invincible > 0 ? 0.5 : 1);
      g.drawCircle(this.pos.x, this.pos.y, this.drawRadius);
      g.endFill();
      // line to show facing
      g.lineStyle(1 / Camera.pixelsPerUnit, 0x000000, 1);
      g.moveTo(this.pos.x, this.pos.y);
      const end = this.pos.add(new Vector(this.angle).scale(this.drawRadius));
      g.lineTo(end.x, end.y);
    }
  }
}

File: src/game/creature/Creature.java
package game.creature;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.glfw.GLFW.*;
import game.DebugMode;
import game.Vector;
import game.level.Level;
import game.Input;
import game.Texture;

public class Creature {
	public Vector pos, vel = Vector.zeroVector;
	public float angle;
	
	public float drawRadius = 0.5f;
	public float hitRadius = 0.4f;
	public float mass = 1;
	
	public int health;
	public float invincible = -0.01f;
	public boolean permInvincible = false;
	public boolean solid = true;
	
	public float[] color = null;
	public Texture texture = null;
	
	
	public Creature(int health, Texture texture) {
		this.health = health;
		this.texture = texture;
	}
	
	public Creature(int health, float[] color) {
		this.health = health;
		this.color = color;
	}
	
	
	public void update(float dt, Level level) {
		vel = vel.add(vel.scale(-20*dt));
		
		pos = pos.add(vel.scale(dt));
		
		while (angle >= 2*Math.PI) angle -= 2*Math.PI;
		while (angle <= -2*Math.PI) angle += 2*Math.PI;
		
		if (invincible >= 0)
			invincible -= dt;
		if (permInvincible) invincible = 0.01f;
	}
	
	public void collide(Creature other) {
		if (!solid || !other.solid) return; 
		
		float dist = (other.pos.sub(this.pos)).magnitude();
		float depth = other.hitRadius + this.hitRadius - dist;
		if (depth >= 0) {
			Vector normal = other.pos.sub(this.pos).direction();
			Vector relVel = this.vel.sub(other.vel);
			float relVelR = relVel.dot(normal);
			
			float m1 = this.mass;
			float m2 = other.mass;
			
			float impact = 1/(1/m1 + 1/m2) * relVelR;
			this.vel = this.vel.add(normal.scale(-impact/m1));
			other.vel = other.vel.add(normal.scale(impact/m2));
			
			float disp = 1/(1/m1 + 1/m2) * depth;
			this.pos = this.pos.add(normal.scale(-disp/m1));
			other.pos = other.pos.add(normal.scale(disp/m2));
			
			this.onTouch(other);
			other.onTouch(this);
		}
	}
	
	public void onTouch(Creature other) {}
	
	public void takeHit(int damage, Vector dir) {
		if (invincible > 0) return;
		
		vel = vel.add(dir.direction().scale(1/mass * 20));
		
		if (damage <= 0) return;
		
		health -= damage;
		invincible = 0.7f;
	}
	
	public void render() {
		glPushMatrix();
		glTranslatef(pos.x, pos.y, 0);
		glRotatef((float) (angle*180/Math.PI), 0, 0, 1);
		glScalef(drawRadius, drawRadius, 1);
		
		if (texture == null) {
			if (invincible > 0)
				glColor4f(color[0], color[1], color[2], 0.5f);
			else
				glColor3f(color[0], color[1], color[2]);
			
			glBegin(GL_TRIANGLE_FAN);
			glVertex2f(0, 0);
			for (float i = 0; i < 32; i += 1) {
				float a = (float) (i / 32 * 2*Math.PI);
				glVertex2f((float) Math.cos(a), (float) Math.sin(a));
			}
			glVertex2f(1, 0);
			glEnd();
			
			glColor3f(0, 0, 0);
			glBegin(GL_LINES);
			glVertex2f(0, 0);
			glVertex2f(1, 0);
			glEnd();
                } else {
                        texture.bind();
                        glColor4f(1, 1, 1, invincible > 0 ? 0.5f : 1);
                        glBegin(GL_QUADS);
                        glTexCoord2f(0, 1);
                        glVertex2f(-1, -1);
                        glTexCoord2f(1, 1);
                        glVertex2f(1, -1);
                        glTexCoord2f(1, 0);
                        glVertex2f(1, 1);
                        glTexCoord2f(0, 0);
                        glVertex2f(-1, 1);
                        glEnd();
                        Texture.bindNone();
                }
		
		glPopMatrix();
		
                if (DebugMode.enabled && Input.isKeyDown(GLFW_KEY_6)) {
			glPushMatrix();
			glTranslatef(pos.x, pos.y, 0);
			glRotatef((float) (angle*180/Math.PI), 0, 0, 1);
			glScalef(hitRadius, hitRadius, 1);
			
			glColor3f(1, 1, 1);
			glBegin(GL_LINE_STRIP);
			for (float i = 0; i < 32; i += 1) {
				float a = (float) (i / 32 * 2*Math.PI);
				glVertex2f((float) Math.cos(a), (float) Math.sin(a));
			}
			glVertex2f(1, 0);
			glEnd();
			glPopMatrix();
		}
	}
}
